bofire/utils/torch_tools.py

```diff
@@ -147,24 +148,50 @@ def constrained_objective2botorch(
         objective (BotorchConstrainedObjective): The objective that should be transformed.
 
     Returns:
-        Tuple[List[Callable[[Tensor], Tensor]], List[float]]: List of callables that can be used by botorch for setting up the constrained objective, and
-            list of the corresponding botorch eta values.
+        Tuple[List[Callable[[Tensor], Tensor]], List[float], int]: List of callables that can be used by botorch for setting up the constrained objective,
+            list of the corresponding botorch eta values, final index used by the method (to track for categorical variables)
     """
     assert isinstance(
         objective, ConstrainedObjective
     ), "Objective is not a `ConstrainedObjective`."
     if isinstance(objective, MaximizeSigmoidObjective):
-        return [lambda Z: (Z[..., idx] - objective.tp) * -1.0], [
-            1.0 / objective.steepness
-        ]
+        return (
+            [lambda Z: (Z[..., idx] - objective.tp) * -1.0],
+            [1.0 / objective.steepness],
+            idx + 1,
+        )
     elif isinstance(objective, MinimizeSigmoidObjective):
-        return [lambda Z: (Z[..., idx] - objective.tp)], [1.0 / objective.steepness]
+        return (
+            [lambda Z: (Z[..., idx] - objective.tp)],
+            [1.0 / objective.steepness],
+            idx + 1,
+        )
     elif isinstance(objective, TargetObjective):
-        return [
-            lambda Z: (Z[..., idx] - (objective.target_value - objective.tolerance))
-            * -1.0,
-            lambda Z: (Z[..., idx] - (objective.target_value + objective.tolerance)),
-        ], [1.0 / objective.steepness, 1.0 / objective.steepness]
+        return (
+            [
+                lambda Z: (Z[..., idx] - (objective.target_value - objective.tolerance))
+                * -1.0,
+                lambda Z: (
+                    Z[..., idx] - (objective.target_value + objective.tolerance)
+                ),
+            ],
+            [1.0 / objective.steepness, 1.0 / objective.steepness],
+            idx + 1,
+        )
+    elif isinstance(objective, CategoricalObjective):
+        # The output of a categorical objective has final dim `c` where `c` is number of classes
+        return (
+            [
+                lambda Z: -1.0
+                * objective.w
```

Comment:
Why `objective.w`?

Reply:
This is an error. I was testing things locally; it will be changed in the next revision :)